File: proxy/typescript/Proxy.ts
// Subject interface
interface Image {
  display(): string;
}

// Real Subject
class RealImage implements Image {
  private filename: string;

  constructor(filename: string) {
    this.filename = filename;
    this.loadFromDisk();
  }

  private loadFromDisk(): void {
    console.log(`Loading image from disk: ${this.filename}`);
  }

  display(): string {
    return `Displaying image: ${this.filename}`;
  }
}

// Proxy
class ProxyImage implements Image {
  private filename: string;
  private realImage: RealImage | null = null;

  constructor(filename: string) {
    this.filename = filename;
  }

  display(): string {
    if (this.realImage === null) {
      this.realImage = new RealImage(this.filename);
    }
    return this.realImage.display();
  }
}

// Usage
const image1 = new ProxyImage('photo1.jpg');
const image2 = new ProxyImage('photo2.jpg');

console.log('\nFirst display call:');
console.log(image1.display());

console.log('\nSecond display call (cached):');
console.log(image1.display());

console.log('\nDisplay another image:');
console.log(image2.display());

File: patterns/proxy/python/Proxy.py
from abc import ABC, abstractmethod

# Subject interface
class Image(ABC):
    @abstractmethod
    def display(self) -> str:
        pass

# Real Subject
class RealImage(Image):
    def __init__(self, filename: str):
        self.filename = filename
        self._load_from_disk()

    def _load_from_disk(self):
        print(f"Loading image from disk: {self.filename}")

    def display(self) -> str:
        return f"Displaying image: {self.filename}"

# Proxy
class ProxyImage(Image):
    def __init__(self, filename: str):
        self.filename = filename
        self._real_image: RealImage = None

    def display(self) -> str:
        if self._real_image is None:
            self._real_image = RealImage(self.filename)
        return self._real_image.display()


if __name__ == "__main__":
    # Images are loaded only when displayed
    image1 = ProxyImage("photo1.jpg")
    image2 = ProxyImage("photo2.jpg")

    print("\nFirst display call:")
    print(image1.display())

    print("\nSecond display call (cached):")
    print(image1.display())

    print("\nDisplay another image:")
    print(image2.display())
